### PR TITLE
DD-883 + DD-919: phase 2 - License mapping - publicationDate

### PR DESCRIPTION
**Pull request by jo-pol**
Fixes DD-883 + DD-919: phase 2 - License mapping - publicationDate

# Description of changes

* expected_dataset: column license is split into license_url and license_name
* The ...LoadImpl classes were introduced to allow unit testing but it appears they are not required (at first it was a workaround for transaction problems)
* processing fedora files is made optional
* iterator treated as such
* use publicationDate for dataset metadata by load-from-dataverse

# How to test

When deploying over an existing DB, the old column license will stay.

# Related PRs
(Add links)
*

# Notify
@DANS-KNAW/dataversedans

### CODE REVIEW

File: src/main/java/nl/knaw/dans/migration/core/ExpectedLoader.java [hunk 1]
```diff
@@ -79,15 +82,23 @@ public void retriedSave(ExpectedFile expected) {
...
-      log.trace(expected.toString());
-      expectedDatasetDAO.create(expected);
+    String depositor = expected.getDepositor();
+    expected.setDepositor(userToEmail.getOrDefault(depositor, depositor));
+
+    if (null != expected.getLicenseUrl()) {
+      String url = expected.getLicenseUrl()
+              .replace("https://", "http://")
+              .replaceAll("(.html|.txt|.pdf)?/$", "");
+      expected.setLicenseName(licensesUrlToName.getOrDefault(url, null));
```
Comment (jo-pol) [inline src/main/java/nl/knaw/dans/migration/core/ExpectedLoader.java]: @lindareijnhoudt: Please verify these replacements cover all variations found in real data.

File: src/main/assembly/dist/cfg/licenses.csv [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+http://www.mozilla.org/en-US/MPL/2.0/FAQ,MPL-2.0
+http://www.ohwr.org/attachments/735/CERNOHLv1_1,CERN-OHL-1.1
+http://www.ohwr.org/attachments/2388/cern_ohl_v_1_2,CERN-OHL-1.2
+http://www.ohwr.org/projects/cernohl/wiki,CERN-OHL-1.2
+http://www.tapr.org/ohl,TAPR-OHL-1.0
+http://www.tapr.org/TAPR_Open_Hardware_License_v1.0,TAPR-OHL-1.0
+http://dans.knaw.nl/en/about/organisation-and-policy/legal-information/DANSLicence,DANS_Licence_UK
+http://www.cecill.info/licences/Licence_CeCILL_V2-en,CeCILL_V2
+http://www.cecill.info/licences/Licence_CeCILL-B_V1-en,CeCILL-B_V1
+http://dans.knaw.nl/en/about/organisation-and-policy/legal-information/DANSGeneralconditionsofuseUKDEF,DANS_Licence_UK
```
Comment (jo-pol) [inline src/main/assembly/dist/cfg/licenses.csv]: @lindareijnhoudt please verify the punctuation of the names

File: src/main/java/nl/knaw/dans/migration/core/ExpectedLoader.java (hunk 1 above)
Comment (lindareijnhoudt) [reply src/main/java/nl/knaw/dans/migration/core/ExpectedLoader.java]: i think you can trust that the url used in Fedora will be in the easy-licenses list, as they were taken from them in the first place. So i hope you don't need this kind of replacing

File: src/main/java/nl/knaw/dans/migration/cli/LoadFromFedoraCommand.java [hunk 3]
```diff
@@ -63,7 +64,14 @@ public LoadFromFedoraCommand(
...
     public void configure(Subparser subparser) {
         super.configure(subparser);
-        subparser.addArgument("csv")
+        subparser.addArgument("-f", "--withFiles")
+            .dest(WITH_FILES)
+            .type(Boolean.class)
+            .setConst(true)
+            .setDefault(false)
+            .nargs("?")
+            .help("The table expected_files is not filled without this option");
```
Comment (jo-pol) [inline src/main/java/nl/knaw/dans/migration/cli/LoadFromFedoraCommand.java:73]: The javadoc of `nargs`seems to suggest a configuration for options without a required value, but I'm afraid the first positional argument is consumed as a value for `--withFiles`. So you have to specify `true`.

File: docs/index.md [hunk 4]
```diff
@@ -58,6 +58,8 @@ ARGUMENTS
 
       named arguments:
         -c [FILE], --config [FILE]   application configuration file (default: etc/config.yml)
+        -f [{true,false}], --withFiles [{true,false}]
+                                     The table expected_files is not filled without this option (default: false)
```
Comment (lindareijnhoudt) [inline docs/index.md:62]:
```suggestion
                                     The table expected_files is only filled with this option specified (default: false)
```

File: src/main/java/nl/knaw/dans/migration/core/DataverseLoader.java [hunk 5]
```diff
@@ -48,11 +50,13 @@ public DataverseLoader(DataverseClient client, ActualFileDAO actualFileDAO, Actu
         this.actualDatasetDAO = actualDatasetDAO;
     }
 
+    @UnitOfWork("hibernate")
```
Comment (lindareijnhoudt) [inline src/main/java/nl/knaw/dans/migration/core/DataverseLoader.java:53]: is this a transactional annotation? Does this mean every create is a separate call to the db?

File: src/test/resources/debug-etc/licenses.csv [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+http://www.gnu.org/licenses/gpl-3.0.en,GPL-3.0
+http://www.gnu.org/licenses/lgpl-3.0,LGPL-3.0
+http://www.gnu.org/licenses/old-licenses/gpl-2.0.en,GPL-2.0
+http://www.mozilla.org/en-US/MPL/2.0/FAQ,MPL-2.0
+http://www.ohwr.org/attachments/735/CERNOHLv1_1,CERN-OHL-1.1
+http://www.ohwr.org/attachments/2388/cern_ohl_v_1_2,CERN-OHL-1.2
+http://www.ohwr.org/projects/cernohl/wiki,CERN-OHL-1.2
+http://www.tapr.org/ohl,TAPR-OHL-1.0
+http://www.tapr.org/TAPR_Open_Hardware_License_v1.0,TAPR-OHL-1.0
+http://dans.knaw.nl/en/about/organisation-and-policy/legal-information/DANSLicence,DANS_Licence_UK
```
Comment (lindareijnhoudt) [inline src/test/resources/debug-etc/licenses.csv:26]:
```suggestion
https://dans.knaw.nl/en/about/organisation-and-policy/legal-information/DANSLicence,DANS_Licence_UK
```

**Pull request merged**